### ISSUE TITLE
Doesn't work in Chrome 55.0.2883.87 m (64-bit)

### ISSUE THREAD

**Issue by davidderoo** [altert/OpenseadragonFabricjsOverlay#11]
Hi,
Since chrome updated to 55.0.2883.87 m (64-bit) it can't select the objects any more.
Even with the demo you have online.
The demos on fabricjs website are working well.

Kind regards,

david de roo

**Comment by MBuLoft1** [comment]
Same here. Chrome 55.0.2883.87 (64-bit) on Mac.
I noticed that a click on a FabricJS element never triggers the 'mouse:down' event in the plugin code, while it does on Firefox and Safari.
```
        this._fabricCanvas.on('mouse:down', function (options) {
            if (options.target) {
                options.e.preventDefault();
                options.e.stopPropagation();
            }
        });
```

**Comment by gd-theli** [comment]
I also encountered the same problem...

**Comment by MBuLoft1** [comment]
(note: by the way, the same problem as reported here is also in Chrome Beta 56)

FYI, I commented the following lines in openseadragon.js and this serves as a workaround in Chrome 55, at least in my application the selection of fabricjs elements is working now. I am unsure yet as to what side effect this may or may not have.
```
            pointerover:           function ( event ) { onPointerOver( _this, event ); },
            MSPointerOver:         function ( event ) { onPointerOver( _this, event ); },
            pointerout:            function ( event ) { onPointerOut( _this, event ); },
            MSPointerOut:          function ( event ) { onPointerOut( _this, event ); },
            pointerdown:           function ( event ) { onPointerDown( _this, event ); },
            MSPointerDown:         function ( event ) { onPointerDown( _this, event ); },
            pointerup:             function ( event ) { onPointerUp( _this, event ); },
            MSPointerUp:           function ( event ) { onPointerUp( _this, event ); },
            pointermove:           function ( event ) { onPointerMove( _this, event ); },
            MSPointerMove:         function ( event ) { onPointerMove( _this, event ); },
            pointercancel:         function ( event ) { onPointerCancel( _this, event ); },
            MSPointerCancel:       function ( event ) { onPointerCancel( _this, event ); },
            pointerupcaptured:     function ( event ) { onPointerUpCaptured( _this, event ); },
            pointermovecaptured:   function ( event ) { onPointerMoveCaptured( _this, event ); },
```
If anyone can suggest a clean way to accomplish the same, please let us know.

**Comment by MBuLoft1** [comment]
FYI, a better workaround is to simply disable pointer events as follows in Openseadragon (openseadragon.js) 
```
// if ( window.PointerEvent && ( window.navigator.pointerEnabled || $.Browser.vendor !== $.BROWSERS.IE ) ) {
if(false) {
```

This does not seem to have side effects. It is still a workaround but useful for anyone looking for a quick fix. Better suggestions are welcome...

**Comment by iangilman** [comment]
@msalsbery Heads up... looks like the new Chrome has some changes that are causing issues with OSD pointer event handling. Any thoughts?

**Comment by msalsbery** [comment]
@iangilman Thank you...there seems to be a bunch of issues elsewhere (FF) as well. I have to figure out how to attack this...

**Comment by iangilman** [comment]
@msalsbery Sounds good... would be great to get it sorted out!

Maybe a first step would be to file an issue in OSD enumerating what the problems are?

**Comment by Faibk** [comment]
Before Chrome Version 54 the Pointer Event API was considered experimental. In 54 the API became officially enabled by default instead of sitting behind a flag. That is where the OSD pointer event problems started to occur. They still exist in Chrome 59.

A workaround without having to modify OSD code is just to disable the pointer event API here: [chrome://flags/#enable-pointer-events](chrome://flags/#enable-pointer-events). Be aware, that this might affect your application as well if you use the API. In my case it works fine and I can safely update the OSD version.

**Comment by iangilman** [comment]
@Faibk Good to know! That only works for your local copy of Chrome, though, right? Is that enough in your case? 

Any thoughts about a more comprehensive fix?

**Comment by Faibk** [comment]
Yes, this is only local. In my case this is enough, because my app is only used in a defined environment where I control the browsers. As you said this would not be applicable for server-deployed applications. 

Since my use case allows it, I opted for the browser flag, because it resolves the issue of editing an external dependecy. I would need to make the changes every time I update to a newer OSD Version.

Approaching a more comprehensive fix, I can only say that it is not the Pointer Event API per se that causes problems, but the merge of touch and click events it introduced. I got the impression while debugging that the events were not correctly handled and handed over to the fabricjs code. As @MBuLoft1 wrote earlier the `mouse:down` event listeners are not even called. In my tests the corresponding touch events appeared instead. If I have time I will look into the OSD Code, I just came in touch with OSD last week when I took over some code from a colleague. 
I also saw, that OSD made some changes to the touch and pointer events in the last release, maybe we can try this together with the FabricJs Overlay here.

**Comment by altert** [comment]
@Faibk Very interesting! what touch event appeared in fabric js? I've tried to catch dragging touch events, but they don't seem to fire either

**Comment by Faibk** [comment]
I could not reproduce the touch event I thought I got while debugging, so I'm not sure if this was related to something I did in the code last time. But I found the overall bug if you can even call it one. Fabricjs simply does not support pointer events. So when OSD discovers pointer support and stops dispatching mouse events in favor of pointer events, fabricjs will just never fire a listener. That is why disabling pointer event detection in OSD in some kind worked as a workaround.

I could fix all issues mentioned in this thread by just replacing following events in the fabric.js:
mousedown -> pointerdown
mouseup -> pointerup
mousemove -> pointermove
mouseenter -> pointerenter
mouseout -> pointerout

Looking at a more permanent fix there are three options: 

1. Mapping the pointer events received at the fabricjs overlay canvas (dispatched by the underlying OSD) to corresponding mouse events. Meaning for every received event a new mouse event with the same specs will be dispatched to fabricjs. This could be labor intensive to get right, but would be most flexible for upgrading this overlay repository to new fabricjs versions.

2. Stick to the current workflow of fully including fabricjs in the repo and just make a comment somewhere to replace the events every time a new fabric version is pulled into the repo.

3. Eventually fabricjs will integrate Pointer Events since it is supposed to be the future of input devices in browsers. So we could open an issue at their github and wait for it.

Probably Option 2 paired with 3 would be the way to go to keep the workload small. 

Obviously the replacement of all mouse events with pointer events needs to be tested a bit more extensively before it should be used.

**Comment by Faibk** [comment]
Oh and I just realized, this will break support for browsers without pointer events... So tomorrow I will try and make something similar to the OSD mechanism, where it only uses pointer events when they are available.

**Comment by altert** [comment]
@Faibk your reasoning for modifying Fabric instead of OSD makes perfect sense. I'd vote for 2 as well. BTW, have you considered actually making a PR to Fabric, or your solution is only temporary?

**Comment by Faibk** [comment]
This depends on how well this fix performs, because I do not know about the internals and complex use cases of fabric. I am not sure if the replacement of mouse events with pointer events is all what needs to be done. I imagine there are some pitfalls, especially with browers only supporting basic touch events. I will create an issue over at fabricjs and ask the experts about it.

**Comment by Faibk** [comment]
Here we go: https://github.com/kangax/fabric.js/issues/4134

**Comment by altert** [comment]
I'm closing this for now. Thanks, @Faibk for fabric version accepting pointer events. It's included in d143a87ad981d14a5964cc502d44852bfbe3bef9 0.2.0 release
